src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1698,18 +1698,152 @@ private static bool CheckNotNamespaceOrType(BoundExpression expr, DiagnosticBag
             }
         }
 
-        private BoundAssignmentOperator BindAssignment(AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
+        private BoundExpression BindAssignment(AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
         {
             Debug.Assert(node != null);
             Debug.Assert(node.Left != null);
             Debug.Assert(node.Right != null);
 
+            if (node.Left.Kind() == SyntaxKind.TupleExpression)
+            {
+                return BindDeconstructionAssignment(node, diagnostics);
+            }
+
             var op1 = BindValue(node.Left, diagnostics, BindValueKind.Assignment); // , BIND_MEMBERSET);
             var op2 = BindValue(node.Right, diagnostics, BindValueKind.RValue); // , BIND_RVALUEREQUIRED);
 
             return BindAssignment(node, op1, op2, diagnostics);
         }
 
+        private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
+        {
+            SeparatedSyntaxList<ArgumentSyntax> arguments = ((TupleExpressionSyntax)node.Left).Arguments;
+
+            // receiver for Deconstruct
+            var boundRHS = BindValue(node.Right, diagnostics, BindValueKind.RValue);
+
+            int numElements = arguments.Count;
+            if (numElements < 2)
+            {
+                // this should be a parse error already.
+                BoundExpression[] args = numElements == 1 ?
+                    new[] { BindValue(arguments[0].Expression, diagnostics, BindValueKind.Assignment), boundRHS } :
+                    new[] { boundRHS };
+
+                return BadExpression(node, args);
+            }
+
+            // bind the variables and check they can be assigned to
+            var checkedVariablesBuilder = ArrayBuilder<BoundExpression>.GetInstance(numElements);
+            for (int i = 0, l = numElements; i < l; i++)
+            {
+                var boundVariable = BindExpression(arguments[i].Expression, diagnostics, invoked: false, indexed: false);
+                var checkedVariable = CheckValue(boundVariable, BindValueKind.Assignment, diagnostics);
+
+                checkedVariablesBuilder.Add(checkedVariable);
+            }
+
+            var checkedVariables = checkedVariablesBuilder.ToImmutableAndFree();
+
+            // symbol and parameters for Deconstruct
+            MethodSymbol deconstructMethod;
+            ImmutableArray<TypeSymbol> parameterTypes;
+            if (!FindDeconstruct(checkedVariables, boundRHS, out deconstructMethod, out parameterTypes, node, diagnostics))
```

Comment:
Is it our intent that we not support `(previous, current, next) = (current, next, null)` because the right-hand-side doesn't have a type from which to look up a `Deconstruct` method? Do we have an open issue for that?